### ISSUE TITLE
Admin can delete a User

### ISSUE THREAD

**Issue by sammachin** [flowforge/flowforge#338]
### Description

**As an admin**

**I want to delete a user account**

**So that they are unable to login to flowforge and all their node-red sessions are invalidated**

### Acceptance Criteria

- [ ] User is Removed from list of users
- [ ] Email address *can* be re-used on a new account
- [ ] User is logged out of any Node-RED Editor Sessions
- [ ] Teams & Projects belonging to the user are *not* removed

**Comment by sammachin** [comment]
#342 is for an admin to join a team

**Comment by knolleary** [comment]
I think we need to define what it means for a Team to be left without an Owner. The team will effectively be suspended (#222 ). Would be good to look at how other services handle this operation - eg GitHub

**Comment by sammachin** [comment]
One option would be to say that a team _can't_ be left without an owner, therefore in order to delete a user that is an owner of a team that team must have a 2nd owner first. Which could be that the admin adds themselves to the team as owner then removes the user?

**Comment by sammachin** [comment]
Thinking about the scenarios where an admin would want to delete a user account:

1. Within a public service like ff.cloud a user requests removal/closure of their account
2. Within a public service the admin determines that the user should be removed from the platform (eg abuse)
3. Within a corporate/internal service where the user has left the organisation

Scenario 1 is the simplest, if a user requests removal of their account they should be instructed to remove themselves as owner of any teams, either by deleting the team (and projects) or by appointing a new owner and wither removing themselves or stepping down to user. They should also be reminded to update the billing details when transferring ownership.
This requires  delete user button in the admin panel and backend 
The system should check that the user is not an owner of any teams before deletion and fail if they are.

Scenario 2
The admin will need to add a new owner to the team (probably themselves) and then either remove the projects & delete the team or just remove the user as an owner depending on the scenario. If they keep the projects running consideration needs to be given to billing this would be a business decision rather than a technical one, so an adjustment could be made in the billing system.
Again the delete user function of S1 will work here, in addition #342 is needed so that the admin can modify members of the team.

Scenario 3 is basically the same as scenario 1 or 2, a departing employee should ensure that they have handed over ownership of teams before they leave so the admin can then cleanly delete their user, however if not then the admin will need to intervene and appoint new owners before they can remove the user.


In summary the story therefore becomes 

***
As an admin

I want to delete a user account if they are not the owner of any teams

So that they are unable to login to and all their node-red sessions are invalidated
***

**Comment by sammachin** [comment]
can rollover to 0.5

**Comment by knolleary** [comment]
### Backend changes

A new api end point is needed in `forge/routes/api/users.js`:

 - `DELETE /api/v1/users/:id`

All routes in this file are already protected with `verifyAdmin`

The `User` db model should have a `beforeDestroy` hook added. That will enforce the constraints:
 - A user cannot be deleted if they are the only owner of a team
 - A user cannot be deleted if they are the only Admin of the platform
 - ...

(Check the `Team` model for an example of a `beforeDestroy` hook that does this type of check)

Other resources to delete:

 - `Session` instances owned by the user - need to check if the database associations will auto delete these when the user is removed. If they don't, then either need to modify the db model with a migration to setup cascade delete, or add the logic to `beforeDestroy` to do the work.
 - `Invitation` instances that reference the user. Again, need to check if the db will auto delete or not.

### UI Changes

The `admin/users/general` view lists users with a dropdown menu containing the single option to edit a user.

<img width="367" alt="image" src="https://user-images.githubusercontent.com/51083/164231219-fc9da7bf-68be-4029-ab34-93d557e3fcf4.png">

The edit user dialog includes a 'danger zone' section to expire their password. A quick update will be to add a delete user button to this section. That should trigger a secondary confirmation as there is no going back.

That said, the edit user dialog is getting quite cluttered and will need a rework to accommodate any more options. It may be better to move the expire option to the dropdown menu and add 'delete' to that menu as well.

Either route is acceptable for this story.

**Comment by sammachin** [comment]
@knolleary what is the experience of deleting a session while the user is logged into the node-red editor? 
Am I right in thinking it will continue but then fail when the try a deploy?

**Comment by knolleary** [comment]
That is something we need to review generally (see also #442). NR doesn't revalidate the access token with the core platform on every request - so currently there is no way for the NR session to be remotely ended.

The Storage plugin for NR sends storage requests back to the forge app - however those requests do not contain info about what NR user has triggered the storage event. That is something we want to improve with the new Storage API (which isn't in plan anymore for 3.0... but should be looked at for 3.1).

**Comment by PLCMercenary** [comment]
> @knolleary what is the experience of deleting a session while the user is logged into the node-red editor? Am I right in thinking it will continue but then fail when the try a deploy?

@sammachin @knolleary  This is exactly the experience that occurs when using an external authentication app such as authelia, user is able to operate until they attempt to deploy, and deployment fails, and any of the changes are lost. and user has to re-authenticate.
